vendor/github.com/ugorji/go/codec/helper_unsafe.go

```diff
@@ -0,0 +1,1350 @@
+	// MARKER: inline zerocopystate directly so genHelper forwarding function fits within inlining cost
+
+	// if d.zerocopystate() {
+	if d.decByteState == decByteStateZerocopy && d.h.ZeroCopy {
+		return stringView(v)
+	}
+	return d.string(v)
+}
+
+func (d *Decoder) mapKeyString(callFnRvk *bool, kstrbs, kstr2bs *[]byte) string {
+	if !d.zerocopystate() {
+		*callFnRvk = true
+		if d.decByteState == decByteStateReuseBuf {
+			*kstrbs = append((*kstrbs)[:0], (*kstr2bs)...)
+			*kstr2bs = *kstrbs
+		}
+	}
+	return stringView(*kstr2bs)
+}
+
+// ---------- DECODER optimized ---------------
+
+func (d *Decoder) jsondriver() *jsonDecDriver {
+	return (*jsonDecDriver)((*unsafeIntf)(unsafe.Pointer(&d.d)).ptr)
+}
+
+// ---------- structFieldInfo optimized ---------------
+
+func (n *structFieldInfoPathNode) rvField(v reflect.Value) (rv reflect.Value) {
+	// we already know this is exported, and maybe embedded (based on what si says)
+	uv := (*unsafeReflectValue)(unsafe.Pointer(&v))
+	urv := (*unsafeReflectValue)(unsafe.Pointer(&rv))
+	// clear flagEmbedRO if necessary, and inherit permission bits from v
+	urv.flag = uv.flag&(unsafeFlagStickyRO|unsafeFlagIndir|unsafeFlagAddr) | uintptr(n.kind)
+	urv.typ = ((*unsafeIntf)(unsafe.Pointer(&n.typ))).ptr
+	urv.ptr = unsafe.Pointer(uintptr(uv.ptr) + uintptr(n.offset))
+	return
+}
+
+// runtime chan and map are designed such that the first field is the count.
+// len builtin uses this to get the length of a chan/map easily.
+// leverage this knowledge, since maplen and chanlen functions from runtime package
+// are go:linkname'd here, and thus not inlined as of go1.16beta
+
+func len_map_chan(m unsafe.Pointer) int {
+	if m == nil {
+		return 0
+	}
+	return *((*int)(m))
+}
+
+func len_map(m unsafe.Pointer) int {
+	// return maplen(m)
+	return len_map_chan(m)
+}
+func len_chan(m unsafe.Pointer) int {
+	// return chanlen(m)
+	return len_map_chan(m)
+}
+
+func unsafeNew(typ unsafe.Pointer) unsafe.Pointer {
+	return mallocgc(rtsize2(typ), typ, true)
+}
+
+// ---------- go linknames (LINKED to runtime/reflect) ---------------
+
+// MARKER: always check that these linknames match subsequent versions of go
+//
+// Note that as of Jan 2021 (go 1.16 release), go:linkname(s) are not inlined
+// outside of the standard library use (e.g. within sync, reflect, etc).
+// If these link'ed functions were normally inlined, calling them here would
+// not necessarily give a performance boost, due to function overhead.
+//
+// However, it seems most of these functions are not inlined anyway,
+// as only maplen, chanlen and mapaccess are small enough to get inlined.
+//
+//   We checked this by going into $GOROOT/src/runtime and running:
+//   $ go build -tags codec.notfastpath -gcflags "-m=2"
+
+// reflect.{unsafe_New, unsafe_NewArray} are not supported in gollvm,
+// failing with "error: undefined reference" error.
+// however, runtime.{mallocgc, newarray} are supported, so use that instead.
+
+//go:linkname memmove runtime.memmove
```

Comment:
we should not introduce new uses of go:linkname to private runtime methods that can break across go versions

xref jsoniter problems with this in the past:
* https://github.com/kubernetes-sigs/structured-merge-diff/issues/202
* https://github.com/golang/go/issues/48238#issuecomment-917000973
* https://github.com/golang/go/issues/48238#issuecomment-916331048